[otel-dotnet] log trace correlation doc (#933)
splunk/public-o11y-docs: gdi/get-data-in/application/otel-dotnet/instrumentation/connect-traces-logs.rst

   2023-09-15 15:23:17      -> deployment.environment: Str(dev)
   2023-09-15 15:23:17      -> service.version: Str(1.0.0)
   2023-09-15 15:23:17 ScopeLogs #0
   2023-09-15 15:23:17 ScopeLogs SchemaURL: 
   2023-09-15 15:23:17 InstrumentationScope  
   2023-09-15 15:23:17 LogRecord #0
   2023-09-15 15:23:17 ObservedTimestamp: 2023-09-15 13:23:13.1358363 +0000 UTC
   2023-09-15 15:23:17 Timestamp: 2023-09-15 13:23:13.1358363 +0000 UTC
   2023-09-15 15:23:17 SeverityText: Information
   2023-09-15 15:23:17 SeverityNumber: Info(9)
   2023-09-15 15:23:17 Body: Str(Hello from {activity})
   2023-09-15 15:23:17 Attributes:
   2023-09-15 15:23:17      -> activity: Str(LogWrappingActivity)
   2023-09-15 15:23:17 Trace ID: 17512c0247942df04fb30e6090eacb2c
   2023-09-15 15:23:17 Span ID: dc281b062178e72f
   2023-09-15 15:23:17 Flags: 1

.. _dotnet-traces-logs-requirements-otel:

Check compatibility and requirements
====================================================

.. note:: Automatic log to trace correlation only works for .NET applications. For .NET Framework use manual correlation. See :ref:`manual-trace-logs-correlation-otel`.

> ``Microsoft.Extensions.Logging`` versions ``6.0.0`` and above are supported.

Reviewer: ```suggestion
``Microsoft.Extensions.Logging`` version ``6.0.0`` and higher are supported.
```

.. _dotnet-otel-enable-log-correlation:

Activate log correlation
============================

The instrumentation turns on automatic log to trace correlation by default.

If you need to turn off logs enrichment and export, set any of the following environment variables to ``false``:

- ``OTEL_DOTNET_AUTO_LOGS_ENABLED``
- ``OTEL_DOTNET_AUTO_LOGS_INSTRUMENTATION_ENABLED``

> - ``OTEL_DOTNET_AUTO_LOGS_ILOGGER_INSTRUMENTATION_ENABLED``

Reviewer: @lachmatt List is correct? Was it the three of them?

Reviewer 2: Any of them.

Reviewer: I'd go for just one to avoid confusion though. Or reword so that it's clear that any will do.

.. _dotnet-otel-include-trace-data:

Trace metadata in log statements
===================================================

> Individual log records, if exported inside scope of an active activity, will contain trace context information:

Reviewer: ```suggestion
Individual log records, if exported inside the scope of an active activity, contain the following trace context information:
```

* ``TraceId``
* ``SpanId``
* ``TraceFlags``

> :new-page:`OpenTelemetry resource <https://github.com/open-telemetry/opentelemetry-specification/tree/4f23dce407b6fcaba34a049df7c3d41cdd58cb77/specification/resource/semantic_conventions>`

Reviewer: ```suggestion
For more information on entities that can produce telemetry, like logs, see :new-page:`OpenTelemetry resource <https://github.com/open-telemetry/opentelemetry-specification/tree/4f23dce407b6fcaba34a049df7c3d41cdd58cb77/specification/resource/semantic_conventions>`.
```

Author: I prefer the current wording, but will let others express their opinion.

Reviewer 3: alternative wording:

```md
Logs produced by an application always have the following set of OpenTelemetry [resource attributes](https://github.com/open-telemetry/opentelemetry-specification/tree/4f23dce407b6fcaba34a049df7c3d41cdd58cb77/specification/resource/semantic_conventions):```

Reviewer: @Kielek Could you make that a GH suggestion, so that @lachmatt or myself can apply it?

Author: changed in https://github.com/splunk/public-o11y-docs/pull/933/commits/6dda38615f5d105cc462675b50c52398dc6bc4a6

> describes entity producing telemetry, e.g logs.

Reviewer: ```suggestion
```

Author: related to https://github.com/splunk/public-o11y-docs/pull/933#discussion_r1328469907

> Logs produced by application have constant set of OpenTelemetry resource attributes always added to them:

Reviewer: ```suggestion
Logs produced by an application have always have the following set of OpenTelemetry resource attributes:
```

Author: ```suggestion
Logs produced by an application always have the following set of OpenTelemetry resource attributes:
```

* ``telemetry.sdk.name``` with constant value of ``opentelemetry``
* ``telemetry.sdk.language`` with constant value of ``dotnet``
* ``telemetry.sdk.version`` with current version of ``OpenTelemetry .NET SDK``
* ``telemetry.auto.version`` with current version of ``OpenTelemetry .NET instrumentation``
* ``splunk.distro.version`` with current version of ``Splunk Distribution of OpenTelemetry .NET``

Additionally, the instrumentation adds the following attributes:

> * ``service.name``, as defined by the ``OTEL_SERVICE_NAME`` environment variable
> * any attribute configured with ``OTEL_RESOURCE_ATTRIBUTES`` environment variable
> * any attribute added by configured resource detectors (e.g ``container.id``)

Reviewer: ```suggestion
* ``service.name``, as defined by the ``OTEL_SERVICE_NAME`` environment variable
* Any attribute configured using the ``OTEL_RESOURCE_ATTRIBUTES`` environment variable
* Any attribute added by configured resource detectors, like ``container.id``
```

.. _manual-trace-logs-correlation-otel:

Manual log to trace correlation
===================================================

You can configure logging libraries to include tracing attributes in logs written to existing logs destination.

``NLog``
----------------------------------------------------

> You can use :new-page:`NLog.DiagnosticSource <https://www.nuget.org/packages/NLog.DiagnosticSource>`,
> see :new-page:`documentation <https://github.com/NLog/NLog.DiagnosticSource>`.

Reviewer: ```suggestion
You can use :new-page:`NLog.DiagnosticSource <https://www.nuget.org/packages/NLog.DiagnosticSource>`. See the :new-page:`NLog official documentation <https://github.com/NLog/NLog.DiagnosticSource>` for more information.
```

``Serilog``
-----------------------------------------------

> You can use one of available enrichers, e.g :new-page:`Serilog.Enrichers.Span <https://www.nuget.org/packages/Serilog.Enrichers.Span>`
> or create your own :new-page:`enricher <https://github.com/serilog/serilog/wiki/Enrichment>`
> to add trace context as properties to log events.

Reviewer: ```suggestion
You can use one of the available enrichers, like :new-page:`Serilog.Enrichers.Span <https://www.nuget.org/packages/Serilog.Enrichers.Span>` or create your own :new-page:`enricher <https://github.com/serilog/serilog/wiki/Enrichment>`
to add trace context as properties to log events.
```
